cpp/src/kmeans/common.cuh

```diff
@@ -306,26 +441,63 @@ void minClusterDistance(const cumlHandle_impl &handle,
     auto minClusterDistanceView =
       minClusterDistance.template view<1>({ns}, {dIdx});
 
-    // calculate pairwise distance between cluster centroids and current batch
-    // of input dataset
-    kmeans::detail::pairwiseDistance(handle, datasetView, centroids,
-                                     pairwiseDistance, workspace, metric,
-                                     stream);
-
-    MLCommon::LinAlg::coalescedReduction(
-      minClusterDistanceView.data(), pairwiseDistance.data(),
-      nc,  // leading dimension of pairwiseDistance
-      ns,  // second dimension of pairwiseDistance
-      std::numeric_limits<DataT>::max(), stream, false,
-      [=] __device__(DataT val, int i) {  // MainLambda
-        return val;
-      },
-      [=] __device__(DataT a, DataT b) {  // ReduceLambda
-        return (b < a) ? b : a;
-      },
-      [=] __device__(DataT val) {  // FinalLambda
-        return val;
-      });
+    auto L2NormXView = L2NormX.template view<1>({ns}, {dIdx});
+
+    // tile over the centroids
+    for (auto cIdx = 0; cIdx < n_clusters; cIdx += centroidsBatchSize) {
+      // # of centroids for the current batch
+      auto nc = std::min(centroidsBatchSize, n_clusters - cIdx);
+
+      // centroidsView [nc x n_features] - view representing the current batch
+      // of centroids
+      auto centroidsView =
+        centroids.template view<2>({nc, n_features}, {cIdx, 0});
+
+      if (metric == MLCommon::Distance::EucExpandedL2) {
```

Comment:
code duplication fix after the prim is updated accordingly.

Reply:
Done